### PR TITLE
x86-32 float return for 'Rust' ABI: treat all float types consistently

### PR DESCRIPTION
**Pull request by RalfJung**
This helps with https://github.com/rust-lang/rust/issues/131819: for our own ABI on x86-32, we want to *never* use the float registers. The previous logic only considered F32 and F64, but skipped F16 and F128. So I made the logic just apply to all float types.

try-job: i686-gnu
try-job: i686-gnu-nopt

### CODE REVIEW

**Comment by RalfJung** [comment]
Cc @tgross35

File: compiler/rustc_ty_utils/src/abi.rs [hunk 1]
```diff
@@ -702,30 +701,19 @@ fn fn_abi_adjust_for_abi<'tcx>(
...
-                        arg.make_indirect();
-                        return;
-                    }
-                    Abi::ScalarPair(s1, s2)
-                        if matches!(s1.primitive(), Float(F32 | F64))
-                            || matches!(s2.primitive(), Float(F32 | F64)) =>
-                    {
+                let has_float = matches!(arg.layout.abi, Abi::Scalar(s) if matches!(s.primitive(), Float(_)))
+                    || matches!(arg.layout.abi, Abi::ScalarPair(s1, s2)
+                        if matches!(s1.primitive(), Float(_)) || matches!(s2.primitive(), Float(_)));
```
Comment (workingjubilee) [inline compiler/rustc_ty_utils/src/abi.rs:706]: I... don't find this more readable than the full
```rust
let has_float = match arg.layout.abi {
    pat if cond => true,
    _ => false,
};
```

**Comment by workingjubilee** [comment]
Huh. I hadn't noticed a bunch of target-specific code has escaped `rustc_target`.

That will be fixed.

File: compiler/rustc_ty_utils/src/abi.rs (hunk 1 above)
Comment (workingjubilee) [reply compiler/rustc_ty_utils/src/abi.rs:706]: The thing that is throwing me off is the parens-counting combined with rustfmt's very... interesting choices on how to format this, btw, so I'd also be happy with anything that has slightly less parentheses.

**Comment by workingjubilee** [comment]
@RalfJung just to confirm: This winds up applying to both i586 and i686, yes?

File: compiler/rustc_ty_utils/src/abi.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_ty_utils/src/abi.rs:706]: rustfmt didn't format this, it just skips the line. So I hand-formatted.

**Comment by RalfJung** [comment]
> This winds up applying to both i586 and i686, yes?

I assume so. I didn't change that part of the logic.

File: compiler/rustc_ty_utils/src/abi.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_ty_utils/src/abi.rs]: I pushed a different way of writing this. Now rustfmt does format it and IMO not in a good way, as is often the case with `match`...

Comment (workingjubilee) [reply compiler/rustc_ty_utils/src/abi.rs]: Ahh, that would explain...

I don't _love_ this formatting but it's more tolerable to me, so thank you. I will try to deindent this code a bit in refactorings, then its formatting will matter less.

Comment (workingjubilee) [reply compiler/rustc_ty_utils/src/abi.rs]: e.g. I'd also accept this color for the bikeshed.
kinda can't explain why, tbh.
```rust
                let has_float = matches!(arg.layout.abi, Abi::Scalar(s)
                    if matches!(s.primitive(), Float(_)))
                    || matches!(arg.layout.abi, Abi::ScalarPair(s1, s2)
                        if matches!(s1.primitive(), Float(_))
                        || matches!(s2.primitive(), Float(_)));
```

Comment (RalfJung) [reply compiler/rustc_ty_utils/src/abi.rs]: Uh I find that quite hard to read.^^

Comment (workingjubilee) [reply compiler/rustc_ty_utils/src/abi.rs]: yes, I thought so. 😅

**Comment by tgross35** [comment]
👍 for making this change

File: compiler/rustc_ty_utils/src/abi.rs [hunk 2]
```diff
@@ -695,37 +694,31 @@ fn fn_abi_adjust_for_abi<'tcx>(
...
-                    // below, by returning arguments up to the size of a pointer (32 bits on x86)
-                    // cast to an appropriately sized integer.
-                    Abi::Scalar(s) if s.primitive() == Float(F32) => {
-                        // Same size as a pointer, return in a register.
-                        arg.cast_to(Reg::i32());
-                        return;
+                let has_float = match arg.layout.abi {
+                    Abi::Scalar(s) => matches!(s.primitive(), Float(_)),
+                    Abi::ScalarPair(s1, s2) => {
+                        matches!(s1.primitive(), Float(_)) || matches!(s2.primitive(), Float(_))
```
[the remark this answers is not in the document]
Comment (workingjubilee) [reply compiler/rustc_ty_utils/src/abi.rs:708]: Yeah, I kinda agree _morally_ with beetrees but not _actually_: this seems fine as-is.

I don't think it's worth spending too much time on this. While I do need to be able to read this code to refactor it, I do intend to bulldoze it.

**Comment by workingjubilee** [comment]
> While I do need to be able to read this code to refactor it, I do intend to bulldoze it.

And since this seems fine-ish, assuming CI doesn't find a reason to take issue with it, and I would hate to make RalfJung rebase this a bunch:

@bors r+

**Comment by matthiaskrgr** [comment]
@bors r-
failed in 
https://github.com/rust-lang/rust/pull/131924#issuecomment-2423908020
i think

**Comment by RalfJung** [comment]
@workingjubilee I had to update the codegen tests, please have a look.

I decided that it's not worth testing the ABI generation for the same return type 20 times, so I made most of these cases match against just the function name, only a few also check the return type to check the ABI. Argument types weren't covered before so I added them as well to ensure they are checked.

**Comment by tgross35** [comment]
Go figure, those tests have hardly been touched since I added them but they just _happen_ to conflict with https://github.com/rust-lang/rust/pull/131211 that landed a couple hours ago :)

Their intent was mostly to test that we lower to the correct llvm calls and not so much ABI, for when I was wiring things up in the compiler without having too much exposed in the library. We probably could use a separate test just for ABI at some point so that check not repeated so many times here, and to cover different signatures and architectures (same for the other float types).

**Comment by workingjubilee** [comment]
note to self, PR review checklist should include:
- if this PR affects codegen, examine the codegen tests under multiple different optimization levels

**Comment by RalfJung** [comment]
Argh that's a painful conflict. :(

**Comment by RalfJung** [comment]
Ah but must of the diff in f128.rs can actually be removed.

File: tests/codegen/float/f16.rs
Comment (workingjubilee) [inline tests/codegen/float/f16.rs:1]: seems fine

File: tests/codegen/float/f128.rs
Comment (workingjubilee) [inline tests/codegen/float/f128.rs:1]: this test feels like it is getting a bit silly, but otherwise seems fine.

**Review by workingjubilee — APPROVED** [review]
Thanks!

**Comment by workingjubilee** [comment]
profoundly cursed.

**Comment by RalfJung** [comment]
That's about these tests:
```rust

// The "C" ABI for `f16` and `f128` on x86 has never used the x87 floating point stack. Do some
// basic checks to ensure this remains the case for the "Rust" ABI.

// CHECK-LABEL: return_f16:
#[no_mangle]
pub fn return_f16(x: f16) -> f16 {
    // CHECK: pinsrw $0, {{.*}}(%ebp), %xmm0
    // CHECK-NOT: xmm0
    // CHECK: retl
    x
}

// CHECK-LABEL: return_f128:
#[no_mangle]
pub fn return_f128(x: f128) -> f128 {
    // CHECK: movl [[#%d,OFFSET:]](%ebp), %[[PTR:.*]]
    // CHECK-NEXT: movl [[#%d,OFFSET+4]](%ebp), %[[VAL1:.*]]
    // CHECK-NEXT: movl [[#%d,OFFSET+8]](%ebp), %[[VAL2:.*]]
    // CHECK-NEXT: movl [[#%d,OFFSET+12]](%ebp), %[[VAL3:.*]]
    // CHECK-NEXT: movl [[#%d,OFFSET+16]](%ebp), %[[VAL4:.*]]
    // CHECK-NEXT: movl %[[VAL4:.*]] 12(%[[PTR]])
    // CHECK-NEXT: movl %[[VAL3:.*]] 8(%[[PTR]])
    // CHECK-NEXT: movl %[[VAL2:.*]] 4(%[[PTR]])
    // CHECK-NEXT: movl %[[VAL1:.*]] (%[[PTR]])
    // CHECK: retl
    x
}
```

We don't actually want the Rust ABI to use the floating point stack here either. So... should the test just be removed? If not, I'd appreciate advice on how to update them, as I can't read the assembly. Cc @beetrees

**Comment by workingjubilee** [comment]
> pinsrw

So, that is saying "insert a 'word' (really, 2 bytes) into a SIMD vector register", which is relevant for x86...
The new instructions are this:
```asm
           pushl %ebp 
           movl %esp, %ebp 
           movzwl 8(%ebp), %eax 
           popl %ebp 
           retl
```
...AT&T syntax...

...anyways, it is:
- push 32-bit value in ebp to stack
- move 32-bit value from esp to ebp
- take the address you get from offsetting ebp by 8 and then from that, move a 16-bit value into eax, zero extending it to 32 bits
- pop 32-bit value from stack to ebp
- return

...isn't that a lot of work...? shouldn't it be just something like
```asm
           movzwl %esp, %eax 
           retl
```
maybe with an offset...? is it because of the frame pointers?

anyway it should be fine to set the test to be
```
// CHECK: pushl [[PUSHED:%{.*}]]
// CHECK: movl %esp, [[PUSHED]]
// CHECK: movzwl 8([[PUSHED]]), %eax
// CHECK: popl [[PUSHED]]
// CHECK-NEXT: retl
```

**Comment by RalfJung** [comment]
That doesn't quite work, I had to remove the `NEXT` (and fix a typo in the `pushl` line, the curly braces there are wrong). Otherwise I get:
```
/home/r/src/rust/rustc.2/tests/assembly/x86-return-float.rs:312:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: retl
                ^
/home/r/src/rust/rustc.2/build/x86_64-unknown-linux-gnu/test/assembly/x86-return-float.normal/x86-return-float.s:631:2: note: 'next' match was here
 retl
 ^
/home/r/src/rust/rustc.2/build/x86_64-unknown-linux-gnu/test/assembly/x86-return-float.normal/x86-return-float.s:629:11: note: previous match ended here
 popl %ebp
          ^
/home/r/src/rust/rustc.2/build/x86_64-unknown-linux-gnu/test/assembly/x86-return-float.normal/x86-return-float.s:630:1: note: non-matching line after previous match is here
 .cfi_def_cfa %esp, 4
^
```

**Comment by RalfJung** [comment]
The relevant part of the asm looks like this:
```
        629:  popl %ebp 
        630:  .cfi_def_cfa %esp, 4 
        631:  retl 
next:312      !~~~  error: match on wrong line
        632: .Lfunc_end20: 
```

**Comment by workingjubilee** [comment]
Ah, apologies, I hadn't actually `CHECK-NEXT`ed it.

**Comment by workingjubilee** [comment]
Yes.

Those are non-effectual directives, but right now I can't remember how to coax rustc or FileCheck into skipping those.

**Comment by workingjubilee** [comment]
r=me with the i686 try-jobs passing.

**Comment by tgross35** [comment]
I guess CI couldn't find anything this time so it just made up a failure

@bors retry

**Comment by workingjubilee** [comment]
real "make up a guy to get made at him" hours

**Comment by SpriteOvO** [comment]
Looks like CI hasn't played enough.

@bors retry

**Pull request merged**